vltt/assignment4/Solution_4.cpp

```diff
@@ -0,0 +1,138 @@
+#include <iostream>
+#include <algorithm>
+#include <iomanip>
+#include <vector>
+#include <map>
+#include <unordered_set>
+#include <set>
+#include <string>
+#include <memory>
+
+#define EXPECT_EQ(expected, actual)                                       \
+  if ((expected) != (actual)) {                                           \
+    std::cout << __FILE__ << ":" << __LINE__ << " Failed: "               \
+              << "Statement (" << #actual << ") is " << actual            \
+              << ", expected " << (expected) << " (" << #expected << ")." \
+              << std::endl;                                               \
+  } else {                                                                \
+    std::cout << __FILE__ << ":" << __LINE__ << " True " << std::endl;  \
+  }
+
+template <class T>
+inline std::ostream& operator<<(std::ostream& os, const std::set<T>& s) {
+	os << "{";
+	for (auto &it : s) {
+		os << it;
+		std::cout << ", ";
+	}
+	os << "}";
+	return os;
+}
+
+
+void make_set(int num_island, std::vector<int> &parent, std::vector<int> &rank) {
+	parent[num_island] = num_island;
+	rank[num_island] = 0;
+}
+
+int find_set(int num_island, std::vector<int> &parent) {
+	if (num_island == parent[num_island])
+		return num_island;
+	return parent[num_island] = find_set(parent[num_island], parent);
+}
+
+void union_sets(int num_island_a, int num_island_b, std::vector<int> &parent, std::vector<int> &rank) {
+	num_island_a = find_set(num_island_a, parent);
+	num_island_b = find_set(num_island_b, parent);
+	if (num_island_a != num_island_b) {
+		if (rank[num_island_a] < rank[num_island_b])
+			std::swap(num_island_a, num_island_b);
+		parent[num_island_b] = num_island_a;
+		if (rank[num_island_a] == rank[num_island_b])
+			++rank[num_island_a];
+	}
+}
+
+//disjoint-set-union (with path compression and union by rank)
```

Comment:
Can you please expand this comment with more information about how this algorithm works?

I know this is union-find, but whoever is reading this code shouldn't need to know what it is to understand what your code is doing :)